### PR TITLE
Add UART for Linux

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

**Comment by floitsch** [comment]
I'm not often programming at the system level. Please take that into account when reviewing.

File: lib/uart_linux.toit [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  Sets the baud rate to the given $new_rate.
+
+  The receiver should be ready to read and write data at the specified
+    baud rate.
+  */
+  baud_rate= new_rate/int:
+    uart_set_baud_rate_ state_.resource new_rate
+
+  /** The current baud rate. */
+  baud_rate:
```
Comment (kasperl) [inline lib/uart_linux.toit:58]:
```suggestion
  baud_rate -> int:
```

File: lib/uart_linux.toit [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  */
+  close:
+    state := state_
+    if not state: return
+    critical_do:
+      state_ = null
+      uart_close_ state.group state.resource
+
+  /**
+  Writes data to the Port.
```
Comment (kasperl) [inline lib/uart_linux.toit:72]: For consistency.

```suggestion
  Writes data to the port.
```

File: lib/uart_linux.toit [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+/**
+Support for Universal asynchronous receiver-transmitter (UART).
+
+UART features asynchronous communication with an external device on two data
+  pins and two optional flow-control pins.
+*/
+
+/**
+The UART Port exposes the hardware features for communicating with an external
```
Comment (kasperl) [inline lib/uart_linux.toit:18]: For consistency:

```suggestion
The UART port exposes the hardware features for communicating with an external
```

File: src/resources/uart_linux.cc [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+    // File descriptors that are intended for subprocesses have the flags cleared.
+    int fd = open(path, O_CLOEXEC | O_RDWR | O_NONBLOCK);
+    if (fd < 0) return -1;
+    if (!isatty(fd)) {
+      // Doesn't seem to be a serial port.
+      close(fd);
+      return -2;
+    }
+
+
```
Comment (kasperl) [inline src/resources/uart_linux.cc:130]: Drop one newline.

File: src/resources/uart_linux.cc [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+    tty.c_cflag |= CREAD | CLOCAL;
+    tty.c_lflag &= ~ICANON;
+    // Disable all echo, erasure, new-line echo. Might not be necessary.
+    tty.c_lflag &= ~ECHO;
+    tty.c_lflag &= ~ECHOE;
+    tty.c_lflag &= ~ECHONL;
+    // Don't interpret INTR, QUIT and SUSP characters.
+    tty.c_lflag &= ~ISIG;
+
+    // Disable special handling of bytes on receive. Just give the raw data:
```
Comment (kasperl) [inline src/resources/uart_linux.cc:151]:
```suggestion
    // Disable special handling of bytes on receive. Just give the raw data.
```

File: src/resources/uart_linux.cc [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    tty.c_iflag &= ~(IGNBRK|BRKINT|PARMRK|ISTRIP|INLCR|IGNCR|ICRNL);
+
+    // Disable any special handling for the output:
```
Comment (kasperl) [inline src/resources/uart_linux.cc:154]:
```suggestion
    // Disable any special handling for the output.
```

File: src/resources/uart_linux.cc [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+    } else {
+      csize = CS8;
+    }
+    tty.c_cflag &= ~CSIZE;
+    tty.c_cflag |= csize;
+
+    if (tcsetattr(fd, TCSANOW, &tty) != 0) goto fail;
+
+    if (tcflush(fd, TCIOFLUSH) != 0) goto fail;
+
```
Comment (kasperl) [inline src/resources/uart_linux.cc:202]: Drop one newline.

File: src/resources/uart_linux.cc [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+  int id = resource_group->create_uart(path, speed, data_bits, stop_bits, parity);
+  if (id == -1) return Primitive::os_error(errno, process);
+  if (id == -2) INVALID_ARGUMENT;
+
+  IntResource* resource = resource_group->register_id(id);
+  // TODO(florian): if we can't allocate the resource, don't we need to
+  // close the uart?
+  // Is there a way to allocate the resource before opening the uart?
+  // The udp_linux.cc does not close the socket either.
+  // I guess we don't really need to worry, as we are on Linux?
```
Comment (kasperl) [inline src/resources/uart_linux.cc:267]: Yeah, we don't need to worry about this.

File: src/resources/uart_linux.cc [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+  if (!resource) MALLOC_FAILED;
+  resource_proxy->set_external_address(resource);
+  return resource_proxy;
+}
+
```
Comment (kasperl) [inline src/resources/uart_linux.cc:272]: Drop one newline.

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: lib/uart_linux.toit [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+import gpio
+import monitor show ResourceState_
+import reader
+import writer
+
+/**
+Support for Universal asynchronous receiver-transmitter (UART).
+
+UART features asynchronous communication with an external device on two data
```
Comment (erikcorry) [inline lib/uart_linux.toit:13]: Maybe measure serial ports here somewhere.

Comment (erikcorry) [reply lib/uart_linux.toit:13]: *mention

**Review by erikcorry — APPROVED** [review]

File: lib/uart_linux.toit (hunk 10 above)
Comment (floitsch) [reply lib/uart_linux.toit:13]: done.

File: lib/uart_linux.toit [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+  */
+  baud_rate= new_rate/int:
+    uart_set_baud_rate_ state_.resource new_rate
+
+  /** The current baud rate. */
+  baud_rate:
+    return uart_get_baud_rate_ state_.resource
+
+  /**
+  Closes this UART port and release all associated resources.
```
[the remark this answers is not in the document]
Comment (floitsch) [reply lib/uart_linux.toit]: done.

**Pull request merged**